optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+We also note, that when instrumental resolution convolution (`TobyFit`) is used,
+the broadening function in `disp2sqw` should only represent some intrinsic physical broadening
+(such as a quasiparticle lifetime).
+Whereas when `TobyFit` is not used,
+the broadening function in `disp2sqw` will be a combination of both intrinsic and instrumental resolution broadening.
+
+In the current Horace implementation, these model functions are used by the `multifit_sqw` function/class
+(which inherits from `multifit`) to fit to data using a built-in Levenberg-Marquardt optimizer. 
+The [new design for PACE](Model_Optimisation_Design.md) proposes instead to use a pair of classes, `OptFunction` and `OptModel`.
+`OptModel` will be more like the current `multifit` (or `tobyfit`) class,
+whilst `OptFunction` is a wrapper around the user supplied model function to handle parameter bindings and bounds currently handled by `multifit`.
+`OptFunction` should therefore have functionality to transform type 2 input to type 1 input
+whilst `OptModel` would expect only type 1 input.
+
+The current implementation of Horace is Matlab-focussed and user defined model functions must be Matlab functions.
+PACE, however, will have a Python version or interface.
+For ease of use, therefore, users *should* be able to define their models using Python functions.
+In addition, many models of inelastic scattering are quite compute-intensive.
+For this reason, PACE should also be able to support user defined models which are embodied in compiled libraries.
+The aim of this design document is to define the interface between these user defined models, in whichever language, and the Horace/PACE fitting infrastructure.
+
+However, for concreteness, *the rest of this document will assume that the new optimisation **framework** will be written in Matlab.*
+The proposed API may need to change should it be decided to implement the optimisation framework in a different language.
+
+
+# <a href="external_interfaces"></a> Matlab interfaces to external languages
+
+One immediate restriction on non-Matlab user-defined model functions is that its *input* arguments must be "plain-old-data" (POD, numeric types or strings) arrays only, and not objects.
+This is because Matlab does not allow its own objects (instances of Matlab-defined classes) to be passed to external libraries.
+(A workaround for this for the Python user interface using a compiled Horace library exists,
+and is described [here](../../python_interface/design/01_pace_python_high_level_discussion.md#interpreter-calls)).
+This means that the second type of function signatures described in the [optimisation design document](Model_Optimisation_Design.md#optimisation-function-object), 
+where Matlab objects (in particular an `sqw` object) are passed as input to the user defined function,
+will not be possible with non-Matlab functions, unless `OptFunction` unwraps this object into a (several) POD array(s).
```

Comment:
`OptFunction` was designed with exactly this in mind. Wrapped functions will need to have their expected input specified (either explicitly or via some sort of function inspection) and `OptFunction` will contain the mechanism required to produce POD arrays when necessary.

Reply:
Sorry I was not clear. I meant that the user-defined function signature: 
> `arrays_objects(a, ..., p)` of which `S_of_Q_w(sqw_obj, p)` is a specific case

Will not be possible for non-Matlab defined `array_objects` or `S_of_Q_w` because `a` or `sqw` cannot be passed from Matlab to an external library.